### PR TITLE
Refactor datacatalog updater to only accept events from SQS

### PR DESCRIPTION
**Pull request by alxarch**
## Background

This is related to #1957 . That issue was closed with a quickfix but this PR addresses the larger issues with datacatalog updater with a more in-depth approach to improve observability and avoid such bugs in the future. 

This PR refactors the datacatalog updater to separate the distinct actions needed 
for source updates and database 'upgrades' during deployments.
It also replaces deferred lambda execution needed for continuation of `gluesync` tasks with SQS messages so we have better observability of each sync without chasing stuff around in logs.

Note: Originally this PR was made against the enterprise version of Panther. To avoid complex backporting, I recreated it with OSS-only changes here.

## Changes

- Separate events for create log tables and sync database
- Use SQS for *all* messages delivered to the datacatalog updater
- Refactor to not use globals

## Testing

- mage test:go
- e2e testing (in progress, some parts need to be checked during upgrades)

### CODE REVIEW

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	for i := range event.Records {
+		s3Event := &event.Records[i]
+		if e := h.HandleS3EventRecord(ctx, s3Event); e != nil {
+			err = multierr.Append(err, e)
+		}
+	}
+	return
+}
+
+func (h *LambdaHandler) HandleS3EventRecord(ctx context.Context, event *events.S3EventRecord) error {
```
Comment (alxarch) [inline internal/log_analysis/datacatalog_updater/datacatalog/s3.go:47]: We should review if we want to only react to 'PutItem` events.

File: deployments/log_analysis.yml [hunk 2]
```diff
@@ -749,7 +758,8 @@ Resources:
           Statement:
             - Effect: Allow
               Action: lambda:InvokeFunction
-              Resource: !Sub arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:panther-datacatalog-updater
+              Resource:
+                - !Sub arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:panther-logtypes-api
```
Comment (kostaspap) [inline deployments/log_analysis.yml:762]: This API doesn't exist

**Comment by alxarch** [comment]
It is used for the `ListAvailableLogTypes` which does exist in OSS as well
AFAIK.
What is not there in OSS are the custom log types endpoints.

**Review by giorgosp — APPROVED** [review]
LGTM! Nice work, please make sure it is tested thoroughly

File: internal/log_analysis/datacatalog_updater/datacatalog/handler.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+type sqsTask struct {
+	Records                []events.S3EventRecord
+	SyncDatabase           *SyncDatabaseEvent
+	CreateTables           *CreateTablesEvent
+	SyncDatabasePartitions *SyncDatabasePartitionsEvent
+	SyncTablePartitions    *SyncTableEvent
+}
+
+func (h *LambdaHandler) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/handler.go:63]: This method implements `lambda.Handler` interface for `datacatalog.LambdaHandler`. This should be the preferred method of defining handlers that should process multiple event types. The reflection based API (i.e. `lambda.Start`) should be avoided. We do not get to control payload unmarshaling that way and we resort to weird tricks for routing our events properly. The `lambda.StartHandler` allows us to control the routing at the `payload` level so we can multiplex events.

File: internal/log_analysis/datacatalog_updater/datacatalog/handler.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	ctx = lambdalogger.Context(ctx, h.Logger)
+	event := events.SQSEvent{}
+	if err := jsoniter.Unmarshal(payload, &event); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal JSON payload")
+	}
+	if err := h.HandleSQSEvent(ctx, &event); err != nil {
+		return nil, err
+	}
+	return []byte(`{"status":"OK"}`), nil
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/handler.go:72]: fix pending

File: internal/log_analysis/datacatalog_updater/datacatalog/create_tables_test.go [hunk 5]
```diff
@@ -57,7 +54,7 @@ func TestSQS_CreateTables(t *testing.T) {
 	// below called once for each database
 	mockGlueClient.On("GetTablesPagesWithContext", mock.Anything, mock.Anything, mock.Anything).Return(nil).Twice()
 	mockAthenaClient := &testutils.AthenaMock{}
-	athenaClient = mockAthenaClient
+	handler.AthenaClient = mockAthenaClient
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/create_tables_test.go:57]: I did not want to have to rewrite all tests. I can invest some more time to have a function returning a handler.

[the remark this answers is not in the document]
Comment (kostaspap) [reply internal/log_analysis/datacatalog_updater/datacatalog/create_tables_test.go:57]: Fair point. If it takes time, you can leave it for now. We can do it on a second pass.

File: internal/log_analysis/awsglue/wrappers.go [hunk 6]
```diff
@@ -157,3 +161,26 @@ func ParseS3URL(s3URL string) (bucket, key string, err error) {
...
+		}
+	}
+	return
+}
+
+func EnsureDatabase(ctx context.Context, client glueiface.GlueAPI, name, description string) error {
+	createDatabaseInput := &glue.CreateDatabaseInput{
+		DatabaseInput: &glue.DatabaseInput{
+			Name:        aws.String(name),
+			Description: aws.String(description),
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/awsglue/wrappers.go:178]: The file already uses `aws.String` extensively. I would prefer to keep the style already set in this file. In such cases where we have a module whose job is to talk to AWS APIs, I think it is best to stick to the `aws.Foo` style. There are cases where it is more readable such as when passing literals:
```
MaxResults: aws.Int64(1),
```
VS
```
maxResults := int64(1)
...
MaxResults maxResults
```
To cover all cases in a unified way I think it is OK to use these functions. As long as this does not add a dependency to `aws` package that is not already there.

File: internal/log_analysis/datacatalog_updater/datacatalog/handler.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+func (h *LambdaHandler) HandleSQSEvent(ctx context.Context, event *events.SQSEvent) error {
+	tasks, err := tasksFromSQSMessages(event.Records...)
+	if err != nil {
+		return err
+	}
+	for _, task := range tasks {
+		switch task := task.(type) {
+		case *events.S3Event:
+			err = h.HandleS3Event(ctx, task)
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/handler.go:83]: The idea of this refactor is that we first 'organize' the SQS event into tasks and then we handle them.
An SQS event can contain multiple `events.S3Event`, each containing multiple `events.S3EventRecord`.
The `tasksFromSQSMessages` function aggregates all the `events.S3EventRecord` into a single `events.S3Event`.

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/aws/aws-lambda-go/events"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/panther-labs/panther/internal/log_analysis/awsglue"
+	"github.com/panther-labs/panther/pkg/lambdalogger"
+)
+
+func (h *LambdaHandler) HandleS3Event(ctx context.Context, event *events.S3Event) (err error) {
+	logger := lambdalogger.FromContext(ctx)
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go:33]: We have to 'leak'. The logger passed in `ctx.Context` is not the same as the logger in `main`. It has added context-related zap fields. This method of 'leaking' is better than the `zap.ReplaceGlobals()` that we used to have before.
The struct is named 'LambdaHandler' anyway :)
If at some point we move to running the handler outside of Lambda, the lambdalogger package would require a rename and a refactor to add context-related information across all environments.

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		return nil
+	}
+	partitionTime := partition.GetTime()
+	tableMeta := partition.GetGlueTableMetadata()
+	if _, err := tableMeta.CreateJSONPartition(h.GlueClient, partitionTime); err != nil {
+		return err
+	}
+	// Store partition in cache as successfully created
+	if h.partitionsCreated == nil {
+		h.partitionsCreated = make(map[string]string)
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go:66]: `map[string]struct{}` requires complex syntax to add elements. `map[string]string` where all keys equal their values has the same space requirements and is much more readable to say
```go
foo["bar"] = "bar"
```
than
```go
foo["bar"] = struct{}{}
```

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kostaspap) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go:33]: This code could as well run in an integration test. I don't see why we really need to run it inside a Lambda. I'd much prefer to used global logger than propagate lambda logic across the code tbh

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kostaspap) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go:66]: We have been using the latter extensively through the code for such cases. Why is this any different? I don't see the readability issue you are referring to. I would prefer to stick to one approach unless we make use of the value.

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go:33]: This is the wrong way to be seeing this. This is not leaking anything lambda related really. Globals are certainly not the answer to any problem and must be avoided at all costs. What we want is a context-aware logger. The only way to get that is by using something similar to `lambdalogger.FromContext`. An integration test without a lambda context would just mute the logging in this case (we can have the function fallback to `zap.L()` if you prefer). If an integration test was testing log output (the only case where this matters anyway) we *should* be passing a lambda context  to verify that the request id is included in the log fields.

File: internal/log_analysis/datacatalog_updater/datacatalog/s3.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/datacatalog_updater/datacatalog/s3.go]: fixed

**Pull request merged**